TribalWars/src/Infrastructure/Repository/BuildingRequiredRepository.cs

```diff
@@ -0,0 +1,69 @@
+﻿using ApplicationCore.Entities;
+using ApplicationCore.Enums;
+using ApplicationCore.Interfaces;
+using Infrastructure.Data;
+using Microsoft.EntityFrameworkCore;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Infrastructure.Repository
+{
+    public class BuildingRequiredRepository : BaseRepository, IBuildingRequiredRepository
+    {
+        public BuildingRequiredRepository(ApplicationDbContext context) : base(context)
+        {
+        }
+
+        public bool HasMaterial(int buildingId, int level, int villageId)
+        {
+            var requiredMaterial = RequiredMaterials(level, buildingId);
+            var playerMaterials = _context.VillageMaterials.Where(material => material.VillageId == villageId).ToList();
+            foreach (var required in requiredMaterial)
+            {
+                var material = playerMaterials.FirstOrDefault(m => m.MaterialId == required.MaterialId);
+                if(material == null || material.Quantity < required.Quantity)
+                {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        public bool HasRequiredBuilding(int buildingId, int villageId)
+        {
+            var requiredBuilding = RequiredBuilding(buildingId);
+            var villageBuilding = _context.VillageBuildings.Where(building => building.VillageId == villageId).ToList();
+            foreach (var required in requiredBuilding)
+            {
+                var building = villageBuilding.FirstOrDefault(b => b.BuildingId == required.BuildingId);
+                if (building == null || required.Level > building.CurrentLevel) 
+                { 
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        public IEnumerable<BuildingRequiredBuilding> RequiredBuilding(int buildingId)
```

Comment:
Usually methods are describing actions, not nouns, so you would rather call these 'GetRequiredBuilding' and 'GetRequiredMaterials'.